From: Christopher Obbard <chris.obbard@collabora.com>
Date: Fri, 22 Oct 2021 17:14:21 +0100
Subject: [HACK] gl-renderer: Take into account viewport when choosing
 filtering

Currently the viewport isn't taken into account when choosing
filtering - we end up with `GL_NEAREST` filtering when we
should have `GL_LINEAR`. Take into account the viewport when
choosing filtering to solve this issue.

Signed-off-by: Christopher Obbard <chris.obbard@collabora.com>
Bug: https://gitlab.freedesktop.org/wayland/weston/-/issues/554
Forwarded: yes
---
 libweston/renderer-gl/gl-renderer.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/libweston/renderer-gl/gl-renderer.c b/libweston/renderer-gl/gl-renderer.c
index d5bfe83..68103bf 100644
--- a/libweston/renderer-gl/gl-renderer.c
+++ b/libweston/renderer-gl/gl-renderer.c
@@ -927,7 +927,9 @@ draw_view(struct weston_view *ev, struct weston_output *output,
 	shader_uniforms(gs->shader, ev, output);
 
 	if (ev->transform.enabled || output->zoom.active ||
-	    output->current_scale != ev->surface->buffer_viewport.buffer.scale)
+	    output->current_scale != ev->surface->buffer_viewport.buffer.scale ||
+		(ev->surface && ev->surface->buffer_viewport.surface.width != -1 &&
+		(ev->surface->buffer_viewport.buffer.src_x + ev->surface->buffer_viewport.buffer.src_width != ev->surface->buffer_viewport.surface.width)))
 		filter = GL_LINEAR;
 	else
 		filter = GL_NEAREST;
